File: tutorials/mpi/examples/ser_heat2D.c
/****************************************************************************
 * FILE: ser_heat2D.c
 * DESCRIPTION:  
 *   Serial HEAT2D Example - C Version
 *   This example is based on a simplified 
 *   two-dimensional heat equation domain decomposition.  The initial 
 *   temperature is computed to be high in the middle of the domain and 
 *   zero at the boundaries.  The boundaries are held at zero throughout 
 *   the simulation.  During the time-stepping, an array containing two 
 *   domains is used; these domains alternate between old data and new data.
 * AUTHOR: D. Turner
 * Last Revised: 04/15/05 Blaise Barney
 ****************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#define NXPROB 100
#define NYPROB 100

struct Parms
{ 
  float cx;
  float cy;
  int nts;
} parms = {0.1, 0.1, 100};

int main(int argc, char *argv[])
{
float u[2][NXPROB][NYPROB];
int ix, iy, iz, it;
void inidat(), prtdat(), update();

printf("Starting serial version of 2D heat example...\n");
printf("Using [%d][%d] grid.\n",NXPROB, NYPROB);

/* Initialize grid and create input file */
printf("Initializing grid and creating input file:");
inidat(NXPROB, NYPROB, u);
prtdat(NXPROB, NYPROB, u, "initial.dat");
for (ix = 0; ix <= NXPROB-1; ix++) {
   u[1][ix][0] = u[0][ix][0];
   u[1][ix][NYPROB-1] = u[0][ix][NYPROB-1];
   }
for (iy = 0; iy <= NYPROB-1; iy++) {
   u[1][0][iy] = u[0][0][iy];
   u[1][NXPROB-1][iy] = u[0][NXPROB-1][iy];
   }

/* Iterate over all timesteps and create output file */
printf("Iterating over %d time steps...\n",parms.nts);
iz = 0;
for (it = 1; it <= parms.nts; it++) {
   update(NXPROB, NYPROB, &u[iz][0][0], &u[1-iz][0][0]);
   iz = 1 - iz;
   }
printf("Done. Created output file: ");
prtdat(NXPROB, NYPROB, &u[iz][0][0], "final.dat");
}


/****************************************************************************
 *  subroutine update
 ****************************************************************************/
void
update(nx, ny, u1, u2)
int nx, ny;
float *u1, *u2;
{
   int ix, iy;

   for (ix = 1; ix <= nx-2; ix++) {
      for (iy = 1; iy <= ny-2; iy++) {
         *(u2+ix*ny+iy) = *(u1+ix*ny+iy)  + 
         parms.cx * (*(u1+(ix+1)*ny+iy) + *(u1+(ix-1)*ny+iy) - 
         2.0 * *(u1+ix*ny+iy)) +
         parms.cy * (*(u1+ix*ny+iy+1) + *(u1+ix*ny+iy-1) - 
         2.0 * *(u1+ix*ny+iy));
         }
      }
}

/*****************************************************************************
 *  subroutine inidat
 *****************************************************************************/
void
inidat(nx, ny, u1)
int nx, ny;
/*float u1[nx][ny];*/
float *u1;
{
   int ix, iy;

   for (ix = 0; ix <= nx-1; ix++) 
      for (iy = 0; iy <= ny-1; iy++) 
         *(u1+ix*ny+iy) = (float)(ix * (nx - ix - 1) * iy * (ny - iy - 1));
}

/**************************************************************************
 * subroutine prtdat
 **************************************************************************/
void
prtdat(nx, ny, u1, fnam)
int nx, ny;
float *u1;
char *fnam;
{
   int ix, iy;
   FILE *fp;

   fp = fopen(fnam, "w");
   for (iy = ny-1; iy >= 0; iy--) {
      for (ix = 0; ix <= nx-1; ix++) {
        fprintf(fp, "%8.3f", *(u1+ix*ny+iy));
        if (ix != nx-1) {
           fprintf(fp, " ");
           }
       else {
          fprintf(fp, "\n");
          }
       }
    }
   fclose(fp);
   printf(" %s\n",fnam);
}

File: mpi/examples/ser_heat2D.f
C ***************************************************************************
C FILE: ser_heat2D.f
C DESCRIPTIONS:  
C   Serial HEAT2D Example - Fortran Version
C   This example is based on a simplified
C   two-dimensional heat equation domain decomposition.  The initial
C   temperature is computed to be high in the middle of the domain and
C   zero at the boundaries.  The boundaries are held at zero throughout
C   the simulation.  During the time-stepping, an array containing two
C   domains is used; these domains alternate between old data and new data.
C AUTHOR: D. Turner
C LAST REVISED: 04/15/05 Blaise Barney
C *************************************************************************

      program heat2D

      parameter (NXPROB = 100, NYPROB = 100)
      dimension u(NXPROB, NYPROB, 2)
      integer nts, tmp1, tmp2
      real cx, cy
      data cx, cy, nts /0.1, 0.1, 100/

      print *, 'Starting serial version of 2D heat example...'
      tmp1 = NXPROB
      tmp2 = NYPROB
      print *, 'Using (',tmp1,',',tmp2,') grid'
 

c *** Initialize grid and create input file 
      print *, 'Initializing grid and creating input file:'
      call inidat(NXPROB, NYPROB, u(1, 1, 1))
      call prtdat(NXPROB, NYPROB, u(1, 1, 1), 'initial.dat')
      do ix = 1, NXPROB
        u(ix, 1, 2) = u(ix, 1, 1)
        u(ix, NYPROB, 2) = u(ix, NYPROB, 1)
      enddo
      do iy = 1, NYPROB
        u(1, iy, 2) = u(1, iy, 1)
        u(NXPROB, iy, 2) = u(NXPROB, iy, 1)
      enddo

c *** Iterate over all timesteps and create output file
      print *, 'Iterating over',nts,' time steps...'
      iz = 1
      do it = 1, nts
        call update(NXPROB, NYPROB, u(1, 1, iz), u(1, 1, 3-iz),cx,cy)
        iz = 3 - iz
      enddo

      print *, 'Done. Created output file:'
      call prtdat(NXPROB, NYPROB, u(1, 1, iz), 'final.dat')
      stop
      end


c****************************************************************************
c*  subroutine update
c****************************************************************************
      subroutine update(nx, ny, u1, u2, cx, cy)
      dimension u1(nx, ny), u2(nx, ny)
      do ix = 2, nx - 1
        do iy = 2, ny - 1
          u2(ix, iy) = u1(ix, iy)  +
     &                 cx * (u1(ix+1,iy) + u1(ix-1,iy) - 2.*u1(ix,iy)) +
     &                 cy * (u1(ix,iy+1) + u1(ix,iy-1) - 2.*u1(ix,iy))
        enddo
      enddo
c
      return
      end
c
c*****************************************************************************
c*  subroutine inidat
c*****************************************************************************/
      subroutine inidat(nx, ny, u1)
      dimension u1(nx, ny)
      do iy = 1, ny
        do ix = 1, nx
          u1(ix, iy) = float((ix - 1) * (nx - ix) * 
     &                       (iy - 1) * (ny - iy)  )
        enddo
      enddo
c
      return
      end


c**************************************************************************
c* subroutine prtdat
c**************************************************************************
      subroutine prtdat(nx, ny, u1, fnam)
      dimension u1(nx, ny)
      character*(*) fnam
      open(17, file=fnam, status='unknown', err=900)
      do iy = ny, 1, -1
        write(17, 1000) (u1(ix, iy), ix = 1, nx)
1000    format(500(f8.3, 1x))
      enddo
      close(17)
      print *, 'Wrote file: ',fnam
      return
900   continue
      print *, 'Error opening file: ', fnam
      stop
      end
